### PR TITLE
Fixed Mistakes (pending test cases)

### PR DESCRIPTION
**Pull request by amnindersingh12**

### CODE REVIEW

File: app/controllers/tweets_controller.rb [hunk 1]
```diff
@@ -53,21 +49,28 @@ def setup_tweet
   def show
     @replies = Tweet.get_replies(params[:id])
     @user = User.find(@tweet.user_id)
+    Visitor.find_or_create_by(tweet_id: @tweet.id, user_id: current_user.id) if current_user.id != @tweet.user.id
```
Comment (swatijadhav) [inline app/controllers/tweets_controller.rb]: @amnindersingh12 
Can we move this to tweet model? and implement like `tweet.add_visitor(current_user)`

File: app/controllers/tweets_controller.rb [hunk 2]
```diff
@@ -53,21 +49,28 @@ def setup_tweet
...
       current_user.like(@tweet)
     end
     respond_to do |format|
-      format.js { render 'likes/like.js' }
+      format.js {}
     end
   end
 
+  def show_likes
+    @likes = Tweet.list_likes(@tweet.id)
```
Comment (swatijadhav) [inline app/controllers/tweets_controller.rb]: it should be `tweet.likes`, make use of associations

File: app/controllers/tweets_controller.rb [hunk 3]
```diff
@@ -53,21 +49,28 @@ def setup_tweet
...
+  end
+
+  def visitors
+    @visitors_are = Tweet.tweet_visitors(@tweet.id)
```
Comment (swatijadhav) [inline app/controllers/tweets_controller.rb]: it should be `tweet.visitors`, make use of associations

File: app/helpers/application_helper.rb [hunk 4]
```diff
@@ -1,4 +1,12 @@
 module ApplicationHelper
+
+  def liked_by(tweet)
+    liked_by = []
+    tweet.likes.each do |user|
+      liked_by << User.find(user.user_id)
+    end
+  end
```
Comment (swatijadhav) [inline app/helpers/application_helper.rb]: This should be a model method, please refer the usage of application helper again.

File: app/views/tweets/_controls.html.erb [hunk 5]
```diff
@@ -22,6 +22,16 @@
         </div>
       <% end %>
     </li>
+    <li>
+      <%= link_to visitor_tweet_path(tweet), class: "text-gray-500 space-x-2 group flex items-center justify-center" do %>
+        <div class="flex items-center justify-center w-10 h-10 rounded-full hover:bg-gray-50">
+          <i class="fa fa-users" aria-hidden="true"></i>
+        </div>
+  <span class="group-hover:text-sky-500 text-gray-500">
+          <%= Tweet.tweet_visitors(tweet).size %>
```
Comment (swatijadhav) [inline app/views/tweets/_controls.html.erb]: it should be simply `tweet.visitors.size`

File: app/views/tweets/like.js.erb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+
+if (<%= @tweet.likes.count >= 1 && @tweet.liked_by?(current_user) %>) {
+$("#like_icon_<%= @tweet.id %>").html('<i class="fa fa-heart" aria-hidden="true"></i>')
+}
+else {
+$("#like_icon_<%= @tweet.id %>").html('<i class="fa fa-heart-o" aria-hidden="true"></i>')
+}
+<%# console.log("hellllo") %>
+<%# <% binding.pry %>
```
Comment (swatijadhav) [inline app/views/tweets/like.js.erb:9]: remove debugger

File: config/routes.rb [hunk 7]
```diff
@@ -6,21 +6,27 @@
   resources :tweets, except: %i[edit update] do
     member do
       post :retweet
-      # get :retweeters
+      get  :visitors, as: :visitor
       post :reply
+      post :like
+      get :show_likes
```
Comment (swatijadhav) [inline config/routes.rb]: use `likes` instead of `show_likes`

**Comment by amnindersingh12** [comment]
Okay @swatijadhav, thank you for the review. I will add the suggested changes to the code

File: config/routes.rb [hunk 8]
```diff
@@ -6,21 +6,27 @@
...
-  post 'like/:id', to: 'tweets#like', as: :like
-  get 'likes', to: 'tweets#like_index', as: :all_users
+  # likes
+  # post 'like/:id', to: 'tweets#like', as: :like
+  # get 'like/:id', to: 'tweets#like_index'
 
-  # resources :likes
-  get "tweets/route", to: 'tweets#route'
+  # visitor counts
+  # get 'visitor/:id', to: 'tweets#visitors', as: :visitor
```
Comment (swatijadhav) [inline config/routes.rb]: remove commented code

File: config/routes.rb [hunk 9]
```diff
@@ -6,21 +6,27 @@
...
+
+  # follow unfollow
   post 'profiles/follow', to: 'profiles#follow'
   delete 'profiles/', to: 'profiles#unfollow'
 
+  # list of followers
   get 'profiles/:id/follower', to: 'profiles#follower', as: :follower
```
Comment (swatijadhav) [inline config/routes.rb]: Move this under profiles resources

File: app/controllers/tweets_controller.rb (hunk 1 above)
Comment (amnindersingh12) [reply app/controllers/tweets_controller.rb]: done

**Review by amnindersingh12 — COMMENTED** [review]
I have made the changes you suggested

File: app/models/tweet.rb [hunk 10]
```diff
@@ -28,7 +28,12 @@ def liked_by?(user)
     likes.where(user:).any?
   end
 
-  has_many :visitors, dependent: :destroy
+  def add_visitor(current_user)
+    Visitor.find_or_create_by(tweet_id: id, user_id: current_user.id)
```
Comment (swatijadhav) [inline app/models/tweet.rb]: @amnindersingh12 check how you can make use of association here

Comment (amnindersingh12) [reply app/models/tweet.rb]: Okay @swatijadhav

File: config/routes.rb [hunk 11]
```diff
@@ -6,28 +6,20 @@
...
-  # follow unfollow
-  post 'profiles/follow', to: 'profiles#follow'
-  delete 'profiles/', to: 'profiles#unfollow'
-
-  # list of followers
-  get 'profiles/:id/follower', to: 'profiles#follower', as: :follower
+  resources :profiles do
+    member do
+      post :follow
+      get :follower
```
Comment (swatijadhav) [inline config/routes.rb:19]: @amnindersingh12 Have you rechecked the follow/unfollow feature after adding this route change?

Comment (amnindersingh12) [reply config/routes.rb:19]: Yes, I did recheck this feature. There was a mistake in my unfollow feature. I have fixed it now, earlier it was 
 `<%= button_to "Unfollow", profiles_path(id: @profile.id), method: :delete`.

**Pull request closed without merging**